io_scene_niftools/modules/nif_import/armature/__init__.py

```diff
@@ -141,33 +142,29 @@ def store_bind_matrices(self, n_armature):
                         if isinstance(geom, NifClasses.BSTriShape):
                             vertex_data = geom.get_vertex_data()
                             if isinstance(geom, NifClasses.BSDynamicTriShape):
-                                # BSDynamicTriShape uses Vector4 to store vertices with a 0 W component, which would
-                                # nullify translation when multiplied by a Matrix44. Hence, first conversion to Vector3
-                                # and assign the position values back later.
-                                vertices = [NifClasses.Vector3.from_value((vertex.x, vertex.y, vertex.z)) for vertex in geom.vertices]
+                                vertices = geom.vertices
                             else:
                                 vertices = [vert_data.vertex for vert_data in vertex_data]
 
                             normals = [vert_data.normal for vert_data in vertex_data]
                         else:
                             vertices = geom.data.vertices
                             normals = geom.data.normals
-                        for vert in vertices:
-                            newvert = vert * diff
-                            vert.x = newvert.x
-                            vert.y = newvert.y
-                            vert.z = newvert.z
-                        diff33 = diff.get_matrix_33()
-                        for norm in normals:
-                            newnorm = norm * diff33
-                            norm.x = newnorm.x
-                            norm.y = newnorm.y
-                            norm.z = newnorm.z
-                        if isinstance(geom, NifClasses.BSDynamicTriShape):
-                            for vertex, t_vertex in zip(geom.vertices, vertices):
-                                vertex.x = t_vertex.x
-                                vertex.y = t_vertex.y
-                                vertex.z = t_vertex.z
+                        # BSDynamicTriShape uses Vector4 to store vertices with a 0 W component, which would
+                        # nullify translation when multiplied by a Matrix44. Hence, first conversion to three-component
+                        # vector
+                        np_vertices = np.array(vertices, dtype=float)[:,:3]
+                        np_vertices = np.concatenate((np_vertices, np.ones((len(np_vertices), 1), dtype=float)), axis=1)
+                        np_diff = np.array(diff.as_list())
+                        np_vertices = np_vertices @ np_diff
+                        np_normals = np.array(normals, dtype=float)
+                        np_diff33 = np.array(diff.get_matrix_33().as_list())
+                        np_normals = np_normals @ np_diff33
+                        # assign the transformed values back
+                        for vertex, t_vertex in zip(vertices, np_vertices[:, :3]):
```

Comment:
or we could just feed np arrays across to the mesh functions to speed things up further. That should save some overhead, too

Reply:
I think for now that's too big a change for me to implement quickly.